grease/src/Grease/Shape.hs

```diff
@@ -314,3 +340,270 @@ deriving instance (ShowExt ext tag, ShowF tag) => Show (ArgShapes ext tag tys)
+                  Just Refl -> Right replace
+                  Nothing ->
+                    Left $
+                      TypeMismatch
+                        { typeMismatchName = nm
+                        , expectedType = C.Some ty
+                        , foundType = C.Some ty'
+                        }
+      Nothing -> Right s
+
+isInSubProg ::
+  -- | the program counter from the base of the target ELF
+  Word64 ->
+  Subprogram ->
+  Bool
+isInSubProg w sub =
+  let entryMatch = (==) w <$> MDwarf.subEntryPC sub
+      def = fromMaybe False
+   in def
+        ( do
+            sdef <- MDwarf.subDef sub
+            lpc <- MDwarf.subLowPC sdef
+            hpc <- MDwarf.subHighPC sdef
+            pure $ w >= lpc && w < (lpc + hpc)
+        )
+        || def entryMatch
+
+rightToMaybe :: Either a b -> Maybe b
+rightToMaybe (Left _) = Nothing
+rightToMaybe (Right b) = Just b
+
+extractType :: Subprogram -> MDwarf.TypeRef -> Maybe MDwarf.TypeApp
+extractType sprog vrTy =
+  do
+    let mp = MDwarf.subTypeMap sprog
+    (mbtypeApp, _) <- Map.lookup vrTy mp
+    rightToMaybe mbtypeApp
+
+constructPtrTarget ::
+  HasPtrWidth w =>
+  TypeUnrollingBound ->
+  Subprogram ->
+  MDwarf.TypeApp ->
+  StateT VisitState Maybe (PtrTarget w NoTag)
+constructPtrTarget tyUnrollBound sprog tyApp =
+  ptrTarget Nothing <$> shapeSeq tyApp
+ where
+  ishape w = lift $ Just $ Seq.singleton $ Initialized NoTag (toBytes w)
+  padding :: (Integral a) => a -> MemShape w NoTag
+  padding w = Uninitialized (toBytes w)
+  -- if we dont know where to place members we have to fail/just place some bytes
+  buildMember :: HasPtrWidth w => Word64 -> MDwarf.Member -> StateT VisitState Maybe (Word64, Seq.Seq (MemShape w NoTag))
+  buildMember loc member =
+    do
+      memLoc <- lift $ MDwarf.memberLoc member
+      padd <- lift $ Just (if memLoc == loc then Seq.empty else Seq.singleton (padding $ memLoc - loc))
+      memberShape <- shapeOfTyApp $ MDwarf.memberType member
+      let memByteSize = sum $ memShapeSize ?ptrWidth <$> memberShape
+      let nextLoc = memLoc + fromIntegral memByteSize
+      lift $ Just (nextLoc, padd Seq.>< memberShape)
+
+  shapeOfTyApp :: (HasPtrWidth w) => MDwarf.TypeRef -> StateT VisitState Maybe (Seq.Seq (MemShape w NoTag))
+  shapeOfTyApp x = shapeSeq =<< lift . extractType sprog =<< pure x
+
+  shapeSeq :: (HasPtrWidth w) => MDwarf.TypeApp -> StateT VisitState Maybe (Seq.Seq (MemShape w NoTag))
+  shapeSeq (MDwarf.UnsignedIntType w) = ishape w
+  shapeSeq (MDwarf.SignedIntType w) = ishape w
+  shapeSeq MDwarf.SignedCharType = ishape (1 :: Int)
+  shapeSeq MDwarf.UnsignedCharType = ishape (1 :: Int)
+  -- Need modification to DWARF to collect DW_TAG_count and properly evaluate dwarf ops for upper bounds in subranges
+  -- https://github.com/GaloisInc/grease/issues/263
+  shapeSeq (MDwarf.ArrayType _elTy _ub) = lift $ Nothing
+  -- need to compute padding between elements, in-front of the first element, and at end
```

Comment:
Just for my own edification: I didn't realize that there could be padding before the first element of a struct. In what situations does that arise?

Reply:
i have no idea if this is possible or exists somewhere. more an artifact of the fact that dwarf would allow such a case to occur simply by declaring no member at offset 0 and we have to do "something"

Reply:
Fair enough, perhaps worth saying that in the comment.